File: Algorithm/Sort/ShellSort/ShellSort.js
const fs = require("fs");

function insertionSort(A, n, h) {
	let num_of_move = 0;

	for (let i = h; i < n; i++) {
		let x = A[i];
		let j = i - h;

		while (j >= 0 && A[j] > x) {
			A[j + h] = A[j];
			j -= h;
			num_of_move++;
		}

		A[j + h] = x;
	}

	console.log(num_of_move);
}

function shellSort(A, n, H) {
	for (let h of H) {
		insertionSort(A, n, h);
	}
}

function main(input) {
	let lines = input.trim().split("\n");
	let n = parseInt(lines[0]);
	let A = lines[1].split(" ").map(Number);
	let k = parseInt(lines[2]);
	let H = lines[3].split(" ").map(Number);

	shellSort(A, n, H);
}

// 標準入力から読み込む
const input = fs.readFileSync("/dev/stdin", "utf8");
main(input);

File: Algorithm/Sort/ShellSort/ShellSort.py
def insertion_sort(A, n, h):
    num_of_move = 0
    for i in range(h, n):
        x = A[i]
        j = i - h
        while j >= 0 and A[j] > x:
            A[j + h] = A[j]
            j -= h
            num_of_move += 1
        A[j + h] = x
    return num_of_move

def shell_sort(A, n, H):
    for h in H:
        num_of_move = insertion_sort(A, n, h)
        print(num_of_move)

# 入力処理
import sys
input = sys.stdin.read
data = input().split()

n = int(data[0])
A = list(map(int, data[1:n+1]))
k = int(data[n+1])
H = list(map(int, data[n+2:n+2+k]))

shell_sort(A, n, H)
